docs/specs/webhook_update.md

```diff
@@ -0,0 +1,87 @@
+## Update webhooks
+
+### Concepts
+
+Instead of having before or after hooks, introduce new concept: blocking and non-blocking hooks.
+
+Actually blocking hooks are the before books and non-blocking is after. But they don't need to be exists in pair anymore.
+
+Blocking hooks accept below response body, operation will be aborted when `is_allowed` is `false`. `title` and `reason` will be shown in auth ui.
+
+```json5
+{
+  "is_allowed": false,
+  "title": "any title",
+  "reason": "any string",
+}
+```
+
+Non-blocking hooks are the same as the after event hooks, response body will be ignored.
+
+
+### Naming convention
+
+For blocking hooks, we create them by different scenarios, so no specific naming convention.
+
+For non-blocking hooks, the main purposes are for data synchronization / app server notification, we will follow the pattern of `<entity>.<action>.<reason>`.
+
+
+### Proposed hooks
+
+#### Blocking hooks
+
+- pre_registration
+- admin_create_user
+
+#### Non-blocking hooks
+
+- user.created.user_signup
+- user.created.admin_api_create
+- identity.created.user_create
+- identity.created.admin_api_create
+- identity.deleted.user_delete
+- identity.deleted.admin_api_delete
+- identity.updated.user_update
+- session.created.user_signup
+- session.created.user_login
+- session.deleted.user_revoke
+- session.deleted.user_logout
+- session.deleted.admin_api_revoke
+- user.promoted.user_promote
+
+### Webhook Event hooks shape
+
+The current webhook shape looks good to me, i want to keep it with some minor update only.
+
+Quick preview of event:
+
+```json5
+{
+  "id": "0E1E9537-DF4F-4AF6-8B48-3DB4574D4F24",
+  "seq": 435,
+  "type": "after_user_create",
+  "payload": { /* ... */ },
+  "context": { /* ... */ }
+}
+```
+
+1. `type` will be the new event type
+1. Add `locale` to context
```

Comment:
Louis's comment:

Let's include more information here

`preferred_languages`: A list of BCP47 tags we inferred from the request
`supported_languages`: The list of supported languages in the configuration.
`fallback_language`: The fallback language in the configuration.
`resolved_language`: The language we resolved(?) Actually it is quite hard to derive this. If the request is made to the web UI. We actually resolved various resources individually. So there could be multiple resolved languages.

The main concern is that if we just pass `preferred_languages`, `supported_languages` and `fallback_language`, the developer could resolve something different from us.